### DISCUSSION TITLE
Question about Ray tracing + Differentiability

### DISCUSSION THREAD

**Discussion by samuelpmish** [shader-slang/slang#5908] [General]
Hi,

I recently learned about slang and it sounds like an incredible language! I'm interested in writing tools to optimize designs of optical systems, so I was excited to read about slang's native support for differentiability and ray tracing. However, glancing at the ray tracing examples, I didn't see any cases where the ray tracing and differentiability were combined. 

Concretely, an example of the problem I'm interested in is: design the shape of some optical surfaces such that image quality is maximized (e.g. simulate refraction through a lens, and vary the vertices of the mesh for the lens to improve sharpness). Is this kind of calculation possible with slang? And if not, is it feasible to help make it possible?

Thank you,

Sam

**Comment by csyonghe** [comment]
The autodiff feature is designed to be orthogonal with most existing graphics or language features. You can can find examples of differentiable path tracers in Falcor: https://github.com/NVIDIAGameWorks/Falcor/tree/master

This frameworks implements an algorithm called warp area sampling to propagate gradients through the ray trace operation to the geometry data.

if you just need to optimize material parameters, then you don’t even need the warp area sampling part, and you just need to make the material evaluation code differentiable.

**Reply by samuelpmish** [reply]
Perfect-- thanks for the information! It looks like there's a wealth of good examples to learn from in Falcor.
